### PR TITLE
Make composite views into a fs2 stream

### PR DESCRIPTION
**Pull request by imsdu**

### CODE REVIEW

File: delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Operation.scala [hunk 1]
```diff
@@ -132,6 +133,15 @@ sealed trait Operation { self =>
 
 object Operation {
 
+  def fromPipe[I: Typeable](elemPipe: ElemPipe[I, Unit]): Operation = new Operation {
```
Comment (imsdu) [inline delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Operation.scala:136]: Will be used later (to save the result of each branch)

File: delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Projection.scala [hunk 2]
```diff
@@ -84,6 +83,33 @@ final class Projection private[stream] (
 
 object Projection {
 
+  private val persistInit: (List[FailedElem], Option[ProjectionProgress]) = (List.empty[FailedElem], None)
+
+  def persist[A](
```
Comment (imsdu) [inline delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Projection.scala:88]: Extracted as a fs2 pipe to be able to use it for composite views

File: delta/plugins/composite-views/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeViewDefSuite.scala [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        projectSource,
+        _ => Right(FilterDeprecated.withConfig(())),
+        (_, run) => toElemStream(run),
+        new NoopSink[NTriples]()
+      )
+      .map { r => (r._1, r._2.name, r._3.name) }
+      .assert((projectSource.id, "main", "rebuild"))
+  }
+
+  test("Compile correctly an Elasticsearch projection") {
```
Comment (olivergrabinski) [inline delta/plugins/composite-views/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeViewDefSuite.scala:57]: Should there be a Blazegraph equivalent of this test?

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeSpaces.scala [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  * @param destroy
+  *   the task to destroy the different namespaces and indices
+  * @param commonSink
+  *   the sink for the common sparql space
+  * @param queryPipe
+  *   the function to create the query pipe from a [[SparqlConstructQuery]]
+  * @param targetSink
+  *   the function to create a sink for a [[CompositeViewProjection]]
+  */
+final case class CompositeSpaces(
```
Comment (imsdu) [inline delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeSpaces.scala:34]: Given the number of namespaces/indices and sinks to instantiate, this allow to separate this logic from the coordinator

File: delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Operation.scala [hunk 5]
```diff
@@ -82,23 +83,48 @@ sealed trait Operation { self =>
...
+      }.flatMap(Stream.chunk)
+  }
+
+  /**
+    * Logs the elements of this stream as they are pulled.
+    *
+    * Logging is not done in `F` because this operation is intended for debugging,
+    * including pure streams.
+    */
+  def debug(formatter: Elem[Out] => String = (elem: Elem[Out]) => elem.toString,
```
Comment (imsdu) [inline delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Operation.scala:97]: This is meant only for debugging purposes and it proved itself useful to do a:
```
operation.debug
```
To display the processed elems at the exit

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeViewDef.scala [hunk 6]
```diff
@@ -0,0 +1,537 @@
...
+    for {
+      pipes        <- source.pipeChain.traverse(compilePipeChain)
+      // We apply `Operation.tap` as we want to keep the GraphResource for the rest of the stream
+      tail         <- Operation.merge(GraphResourceToNTriples, sink).map(_.tap)
+      chain         = pipes.fold(NonEmptyChain.one(tail))(NonEmptyChain(_, tail))
+      operation    <- Operation.merge(chain)
+      // We create the elem stream for the two types of branch
+      // The main source produces an infinite stream and waits for new elements
+      mainSource    = graphStream.main(source, project)
+      // The rebuild one a finite one with obly the current ellements
```
Comment (olivergrabinski) [inline delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeViewDef.scala:469]: Typos

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeViewDef.scala [hunk 7]
```diff
@@ -0,0 +1,537 @@
...
+    * namespace is queried and the result indexed in the associated index/namespace.
+    *
+    * All the source streams are merged to form the main branch.
+    *
+    * ===Rebuild branch===
+    *
+    * When a rebuild strategy is set, an additional stream is built and run following a defined interval when the main
+    * branch has processed new resources.
+    *
+    * A stream that fetches all current resources (that terminates after processing hem) is built for every source.
```
Comment (olivergrabinski) [inline delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeViewDef.scala:95]: them

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeViewDef.scala [hunk 8]
```diff
@@ -0,0 +1,537 @@
...
+    *
+    * It broadcasts immediately the resouces to every projection where it follows the same process as the second part of
```
Comment (olivergrabinski) [inline delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeViewDef.scala:97]: resources

**Review by olivergrabinski — APPROVED** [review]

**Pull request merged**